crates/nu_plugin_parse/src/nu/mod.rs

```diff
@@ -21,114 +23,81 @@ impl Plugin for Parse {
-                                "Could not parse regex",
-                                "could not parse regex",
+                                "Invalid pattern",
+                                "invalid pattern",
                                 tag.span,
                             )
-                        })?;
+                        })?
                     };
+
+                    self.column_names = column_names(&self.regex);
                 }
                 Value { tag, .. } => {
                     return Err(ShellError::labeled_error(
-                        "Unrecognized type in params",
-                        "unexpected value",
+                        format!(
+                            "Unexpected type in params (found `{}`, expected `String`)",
+                            value.type_name()
+                        ),
+                        "unexpected type",
                         tag,
                     ));
                 }
             }
         }
+
         Ok(vec![])
     }
 
     fn filter(&mut self, input: Value) -> Result<Vec<ReturnValue>, ShellError> {
-        match &input.as_string() {
-            Ok(s) => {
-                let mut output = vec![];
-                for caps in self.regex.captures_iter(&s) {
-                    let group_count = caps.len() - 1;
-
-                    if self.column_names.0.len() != group_count {
-                        return Err(ShellError::labeled_error(
-                            format!(
-                                "There are {} column(s) specified in the pattern, but could only match the first {}: [{}]",
-                                self.column_names.0.len(),
-                                group_count,
-                                caps.iter()
-                                    .skip(1)
-                                    .map(|m| {
-                                        if let Some(m) = m {
-                                            let m = m.as_str();
-                                            let mut m = m.replace(",","\\,");
-                                            if m.len() > 20 {
-                                                m.truncate(17);
-                                                format!("{}...", m)
-                                            } else {
-                                                m
-                                            }
-                                        } else {
-                                            "<none>".to_string()
-                                        }
-                                    })
-                                    .collect::<Vec<String>>()
-                                    .join(", ")
-                            ),
-                            "could not match all columns in pattern",
-                            &self.pattern_tag,
-                        ));
-                    }
```

Comment:
It may be good to 👀 this, but I don't think it's ever possible for the column name count and the capture group count here to mismatch. Our simple mode setup will always have all the capture groups (if there was one missing, `captures_iter` could not have made a match).

For `--regex`, there's the potential for an optional capture group, but this is working now given the `zip(captures.iter())` approach I used below:

```
❯ echo ["1:INFO:component:all is well" "2:ERROR::something bad happened"] | parse --regex "(.*?):(.*?):(.*?)?:(.*?)(well)?\z"
───┬──────────┬──────────┬───────────┬────────────────────────┬──────────
 # │ Capture1 │ Capture2 │ Capture3  │ Capture4               │ Capture5
───┼──────────┼──────────┼───────────┼────────────────────────┼──────────
 0 │ 1        │ INFO     │ component │ all is                 │ well
 1 │ 2        │ ERROR    │           │ something bad happened │
───┴──────────┴──────────┴───────────┴────────────────────────┴──────────
```